pkg/client/cache/store.go

```diff
@@ -76,6 +81,57 @@ func (c *cache) Add(obj interface{}) error {
 	c.lock.Lock()
 	defer c.lock.Unlock()
 	c.items[key] = obj
+	c.updateIndices(obj)
```

Comment:
Can you build the list of what you need to modify in the indexes separately from actually making the update?  It would help limit contention on the lock.  Same reason the key is calculated before lock acquisition.